### DISCUSSION TITLE
Rollback rule execution

### DISCUSSION THREAD

**Discussion by Timmy93** [firefly-iii/firefly-iii#8460] [Firefly III] [answered]
Hi everybody, do you know if it is possible to rollback from a wrong rule?
A rule run on 1700+ transaction deleting their category and budget (I don know what happened), is it possible to rollback in any way?
Unfortunately I don't have any backup. Is there any hope?

**Comment by JC5** [comment]
I'm sorry, there isn't :(

**Comment by Timmy93** [comment, marked as answer]
> I'm sorry, there isn't :(

Actually, **I was able to solve the problem**. :) 

Briefly what I did is:
- Extract from the audit log all the changes (_what a golden table, thanks for it!_).
  Here you can find a useful query:
```
; ### Change the dates according to when the rule triggered: `ale.created_at >=` and when you know the rule ended: `ale.created_at <`
SELECT ale.auditable_id, tj.description, action, `before`, `after`
FROM audit_log_entries ale 
JOIN transaction_journals tj on ale.auditable_id = tj.id 
WHERE ale.created_at >= "2024-01-25 16:38:27.000" AND ale.created_at < "2024-01-25 23:38:27.000" 
```
- For each field row extracted, based on the field `action` perform the opposite action to restore the original value using the official REST API
```
### In python is more or less these 3 lines (of course you need to login and set correctly the headers)
api_url = f"{self.url}/api/{self.apiVersion}/transactions/{ff_id}"
data = {
    "transactions": [
        {
            field: restored_value
        }
    ]
}
response = requests.put(api_url, headers=self.headers, json=data)
```

It would be great to have one day this feature in the settings (e.g. rollback to this date)

**Comment by JC5** [comment]
Nice! That is the goal of that table one day, but your trick isn't guaranteed to work (yet), what with deleted budgets for example. But indeed, this should be built in one day. 

Thanks for sharing your solution!

**Reply by Timmy93** [reply]
I don't know if it can be an idea, but for deletion of something (like tags, budget or categories) a long list of audit logs for rollback could be created.
To make a real example the deletion of budget X assigned to transaction #_1, #_2, #_3 could become not a single `action: delete_budget` but could be a rollback of all the previous action done:
- Remove transaction #_3 from budget X
- Remove transaction #_2 from budget X
- Remove transaction #_1 from budget X
- Set budget amount from 1000🪙 to 0🪙
- Set budget name from X to ""

I know, it's not a straightforward implementation and for sure it is not a feature to be used light-heartedly on a daily basis but it would certainly become a lifesaver in the case of "extreme" issues.

Thank you for your helpfulness!

**Reply by JC5** [reply]
That's exactly my plan, yes. Right now the audit data cannot always be trusted so this is tricky, but I hope to make this better in the future.
